### PR TITLE
External contracts package

### PR DESCRIPTION
**Pull request by HenryNguyen5**
https://www.pivotaltracker.com/story/show/170958820

This PR creates two new packages, `@chainlink/linkbelt` and `@chainlink/contracts`. Their readmes are pasted below:

# @chainlink/contracts

Chainlink's solidity contracts and contract abstractions

# Migration Notice

This package stores all of chainlink's smart contracts, previously known as the `evm/` directory. The previous package was called `chainlink`, now it is `@chainlink/core`.

# Layout

This package makes use of `@chainlink/linkbelt` to output artifacts along with ethers.js and truffle contract abstractions from the smart contracts stored within `src`. The resulting layout of the package when published looks like:

```sh
contracts
├── abi # abi output from src/
│   ├── v0.4
│   ├── v0.5
│   └── v0.6
├── ethers # ethers contract abstractions codegenned from abis
│   ├── v0.4
│   ├── v0.5
│   └── v0.6
├── src # the contracts themselves, in .sol form
│   ├── v0.4
│   ├── v0.5
│   └── v0.6
└── truffle  # truffle contract abstractions codegenned from abis
    ├── v0.4
    ├── v0.5
    └── v0.6
```

# Usage

Install this package with the following command:

```sh
yarn add @chainlink/contracts
```

These smart contracts can be imported as a dependency in various ways described below.

## Solidity

The solidity smart contracts themselves can simply be imported via the `src` directory of `@chainlink/contracts`. If you wanted to consume the v0.4.x version of our Oracle smart contract, you could do the following:

```solidity
import "@chainlink/contracts/src/v0.4/Chainlinked.sol";
```

## Artifacts

JSON artifacts generated by [sol-compiler](https://sol-compiler.com/) are available under the `abi` directory. If you wanted to consume the v0.4.x version of our Oracle JSON artifact with Javascript, you could do the following:

```js
const OracleV4Json = require('@chainlink/contracts/abi/v0.4/Oracle.json')
```

## Ethers

This library ships with [ethers.js](https://github.com/ethers-io/ethers.js/) contract abstractions generated by [typechain](https://github.com/ethereum-ts/TypeChain). To use these, make sure you have ethers.js installed as a dependency:

```sh
yarn add ethers@^4.0.44
```

If you wanted to consume the v0.4.x version of our Oracle with Javascript, you could do the following:

```ts
import { OracleFactory } from '@chainlink/contracts/ethers/v0.4/OracleFactory'
```

This gives a fully typed (if using typescript) version of a ethers.js contract factory. See the ethers [documentation](https://docs.ethers.io/ethers.js/html/api-contract.html) on usage.

## Truffle

This library ships with [@truffle/contract](https://github.com/trufflesuite/truffle/tree/master/packages/contract#readme) abstractions of each of our smart contracts. To use these, make sure you have @truffle/contract as a dependency.

```sh
yarn add @truffle/contract@^4.1.8
```

If you wanted to consume the v0.4.x version of our Oracle with Javascript, you could do the following:

```js
const { Oracle } = require('@chainlink/contracts/truffle/v0.4/Oracle')
```

For usage, see the @truffle/contract documentation.

For ease of use with testing, if the environment variable `NODE_ENV` is set, the imported truffle contract abstraction will automatically set its provider to `web3.currentProvider`, avoiding the need to perform a `setProvider` call before using it during your truffle tests.

```sh
NODE_ENV=true yarn truffle test
```

# @chainlink/linkbelt

A toolbelt for performing various commands on chainlink smart contracts.
This cli tool is currently used within `@chainlink/contracts` for the usage of running
build and development tools across multiple solidity contract verisions.

- [Usage](#usage)
- [Commands](#commands)
  

# Usage

```sh-session
$ npm install -g @chainlink/linkbelt
$ linkbelt COMMAND
running command...
$ linkbelt (-v|--version|version)
@chainlink/linkbelt/0.0.1 linux-x64 node-v10.16.3
$ linkbelt --help [COMMAND]
USAGE
  $ linkbelt COMMAND
...
```

# Commands

- [`linkbelt compile [COMPILER]`](#linkbelt-compile-compiler)
- [`linkbelt help [COMMAND]`](#linkbelt-help-command)

## `linkbelt compile [COMPILER]`

Run various compilers and/or codegenners that target solidity smart contracts.

```
USAGE
  $ linkbelt compile [COMPILER]

ARGUMENTS
  COMPILER  (solc|ethers|truffle|all) Compile solidity smart contracts and output their artifacts

OPTIONS
  -c, --config=config  [default: app.config.json] Location of the configuration file
  -h, --help           show CLI help

EXAMPLE
  $ linkbelt compile all

  Creating directory at abi/v0.4...
  Creating directory at abi/v0.5...
  Creating directory at abi/v0.6...
  Compiling 35 contracts...
  ...
  ...
  Aggregator artifact saved!
  AggregatorProxy artifact saved!
  Chainlink artifact saved!
  ...
```

_See code: [src/src/commands/compile.ts](https://github.com/smartcontractkit/chainlink/blob/v0.0.1/src/src/commands/compile.ts)_

## `linkbelt help [COMMAND]`

display help for linkbelt

```
USAGE
  $ linkbelt help [COMMAND]

ARGUMENTS
  COMMAND  command to show help for

OPTIONS
  --all  see all commands in CLI
```

_See code: [@oclif/plugin-help](https://github.com/oclif/plugin-help/blob/v2.2.3/src/commands/help.ts)_

### CODE REVIEW

**Comment by samphilipd** [comment]
This is absolutely enormous. Is there any possible way it can be split into multiple smaller PRs?

File: linkbelt/src/services/config.test.ts [hunk 1]
```diff
@@ -0,0 +1,126 @@
+import mock from 'mock-fs'
```
Comment (se3000) [inline linkbelt/src/services/config.test.ts]: Should this be in the test directory?

File: evm-contracts/package.json [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    "ts-jest": "^25.2.0",
+    "typescript": "^3.7.5"
+  },
+  "files": [
+    "abi/**/*",
+    "ethers/**/*",
+    "src/**/*",
+    "truffle/**/*"
+  ],
+  "optionalDependencies": {
```
[the remark this answers is not in the document]
Comment (HenryNguyen5) [reply evm-contracts/package.json]: Peer dependencies imply that you have to have those dependencies installed for a package to operate. This is not the case for consumers. consumers might only need the ABIs or solidity files that we ship. We should warn consumers during the postinstall step that they need to install either `@truffle/contract` or `ethers` depending on what contract abstraction they want to use, if any.

File: evm-contracts/test/v0.5/Schnorr/Schnorr.test.ts [hunk 3]
```diff
@@ -6,10 +6,11 @@ import {
 } from '@chainlink/test-helpers'
 import { assert } from 'chai'
 import { ethers } from 'ethers'
-import { SchnorrSECP256K1Factory } from '../../src/generated/SchnorrSECP256K1Factory'
+import { SchnorrSECP256K1Factory } from '../../../ethers/v0.5/SchnorrSECP256K1Factory'
 import * as f from './fixtures'
-extensions.ethers.BigNumber.extend()
+
 const { bigNumberify: bn } = ethers.utils
+extensions.ethers.BigNumber.extend(ethers.utils.BigNumber)
```
[the remark this answers is not in the document]
Comment (HenryNguyen5) [reply evm-contracts/test/v0.5/Schnorr/Schnorr.test.ts]: Yeah, let's make that a separate chore, you can accomplish this via https://jestjs.io/docs/en/configuration#globalsetup-string

File: integration-scripts/package.json [hunk 4]
```diff
@@ -28,10 +28,9 @@
     "@0x/sol-compiler": "^4.0.3",
     "@0x/sol-trace": "^3.0.4",
     "@chainlink/test-helpers": "0.0.1",
-    "chainlink": "0.8.2",
-    "chainlinkv0.5": "0.0.3",
+    "@chainlink/contracts": "0.0.1",
```
[the remark this answers is not in the document]
Comment (HenryNguyen5) [reply integration-scripts/package.json]: We can do this after publishing, lets make this a release task

File: linkbelt/src/services/config.test.ts (hunk 1 above)
Comment (HenryNguyen5) [reply linkbelt/src/services/config.test.ts]: I try to keep more unit-ish tests beside the file, and large functional/integration tests outside of the src dir

Comment (se3000) [reply linkbelt/src/services/config.test.ts]: "unit-ish" sounds kind of vague, and this stands out as an outlier atm. Do we do this commonly in other packages? I'd prefer picking a place to look, so people can know where to look when updating and running tests before finding out on CI.

File: integration-scripts/package.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (se3000) [reply integration-scripts/package.json]: https://www.pivotaltracker.com/story/show/171165578

File: linkbelt/src/services/config.test.ts (hunk 1 above)
Comment (HenryNguyen5) [reply linkbelt/src/services/config.test.ts]: We don't really have a standard test guideline.
- Explorer has all tests in `src/__tests__` but it doesnt separate unit and integration tests.
- Explorer-client has unit tests in `src/__tests__` and integration tests in `/integration`
- Operator-ui has unit tests in `__tests__` (not in `src/`) and integration tests in `/integration`
- `@chainlink/contracts` has tests in `tests`

I try to put tests that are unit tests (Not sure why I called this one unit-ish, it's definitely a unit test. I forgot that I mock out the filesystem for this test too.) right beside their files for easily discoverability, and functional/integration tests at top of the package. Happy to change if you think they should be in the top level `tests` folder

File: evm-contracts/test/v0.5/Schnorr/Schnorr.test.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (se3000) [reply evm-contracts/test/v0.5/Schnorr/Schnorr.test.ts]: https://www.pivotaltracker.com/story/show/171165753

File: evm/v0.5/test/support/LinkToken.json [hunk 5]
```diff
@@ -1,163 +0,0 @@
-{
```
[the remark this answers is not in the document]
Comment (se3000) [reply evm/v0.5/test/support/LinkToken.json]: https://www.pivotaltracker.com/story/show/171165757

File: linkbelt/src/services/config.test.ts (hunk 1 above)
Comment (se3000) [reply linkbelt/src/services/config.test.ts]: All the parts that use `integration` test across multiple parts of the repo; `core/`, `operator-ui/`, `explorer`... I think the term "integration" is vague, so if it's within one part of the repo, I lean towards "feature" test. I'm not as familiar with the patterns within the TypeScript packages, but at first glance it still makes sense for me to group them together. Your call.

Comment (HenryNguyen5) [reply linkbelt/src/services/config.test.ts]: I'd vouch for us keeping it as-is and pushing for a guideline for testing in the future to abide by

File: belt/README.md [hunk 6]
```diff
@@ -0,0 +1,87 @@
+# @chainlink/belt
```
Comment (se3000) [inline belt/README.md]: 👨‍⚕ 👩‍⚕

**Pull request merged**